include/irr/asset/filters/CSummedAreaTableImageFilter.h

```diff
@@ -0,0 +1,301 @@
+
+				return true;
+			};
+
+			bool decodeStatus = commonExecute(state, perOutputRegionDecode);
+
+			if (decodeStatus)
+			{
+				const auto arrayLayers = state->inImage->getCreationParameters().arrayLayers;
+				const auto outRegions = state->outImage->getRegions(state->outMipLevel);
+				auto extent = state->outImage->getMipSize(state->outMipLevel);
+				const auto trueOutExtent = asset::VkExtent3D({ extent.x, extent.y, extent.z }); 
+				const auto regionBufferOffset = outRegions.begin()->bufferOffset;
+				auto* imageEntireOutData = reinterpret_cast<uint8_t*>(state->outImage->getBuffer()->getPointer());
+				const size_t singleScratchLayerOffset = (state->extent.width * state->extent.height * state->extent.depth * currentChannelCount); 
+
+				auto sum = [&](core::vectorSIMDu32 readBlockPos) -> void
+				{
+					decltype(readBlockPos) newReadBlockPos = decltype(newReadBlockPos)(readBlockPos.x, trueOutExtent.height - 1 - readBlockPos.y, readBlockPos.z, readBlockPos.w);
+					decodeType* const currentColumnCache = mainColumnCacheScratch + newReadBlockPos.x * currentChannelCount;
+
+					const auto globalIndependentOffset = ((newReadBlockPos.z * state->extent.height + newReadBlockPos.y) * state->extent.width + newReadBlockPos.x);
+					const auto outDataAdressOffsetScratch = globalIndependentOffset * currentChannelCount;
+					const auto layersOffset = singleScratchLayerOffset * newReadBlockPos.w;
+
+					auto finalPixelBuffer = core::make_smart_refctd_ptr<asset::ICPUBuffer>(state->mainRowScratchCacheByteSize);
+					auto finalPixel = reinterpret_cast<decodeType*>(finalPixelBuffer->getPointer());
+					memset(finalPixel, 0, finalPixelBuffer->getSize());
+
+					decodeType* decodeBuffer = scratchMemory + outDataAdressOffsetScratch + layersOffset;
+
+					for (uint8_t channel = 0; channel < currentChannelCount; ++channel)
+					{
+						*(currentColumnCache + channel) += *(mainRowCacheScratch + channel) += decodeBuffer[channel];
+						*(finalPixel + channel) = *(currentColumnCache + channel) - (ExclusiveMode ? decodeBuffer[channel] : 0);
+					}
+
+					memcpy(scratchMemory + outDataAdressOffsetScratch + layersOffset, finalPixel, state->mainRowScratchCacheByteSize);
+
+					if (newReadBlockPos.x == trueOutExtent.width - 1) // reset row cache when changing y
+					{
+						if (newReadBlockPos.y == 0) // sum values in (maxX, 0, z) and remove values from column cache
+						{ 
+							memcpy(totalImageCacheScratch + (newReadBlockPos.w * state->extent.depth + newReadBlockPos.z) * state->mainRowScratchCacheByteSize, finalPixel, state->mainRowScratchCacheByteSize);
+							memset(mainColumnCacheScratch, 0, state->mainColumnScratchCacheByteSize);
+						}
+
+						memset(mainRowCacheScratch, 0, state->mainRowScratchCacheByteSize);
+					}
+				};
+
+				{
+					core::vector3du32_SIMD localCoord;
+					for (auto& w = localCoord[3] = 0u; w < arrayLayers; ++w)
```

Comment:
if you hoist this loop to before decode and have this
```cpp
for (auto& w = localCoord[3] = 0u; w < arrayLayers; ++w)
{
    auto stateCopy = state;
    stateCopy.inBaseLayer += w;
    stateCopy.outBaseLayer += w;
    stateCopy.layerCount = 1u;

    bool decodeStatus = commonExecute(stateCopy, perOutputRegionDecode);
    if (decodeStatus)
    {
        ...
    }
}
```
then you need far less scratch

Reply:
right, I will correct it